File: C++/CppLibrary/ExpressionEvaluator.cpp
#include "ExpressionEvaluator.h"
#include <string>
#include <stdexcept>

ExpressionEvaluator::ExpressionEvaluator(double A, double B, double C, double D)
{
    a = A;
    b = B;
    c = C;
    d = D;
}

void ExpressionEvaluator::SetParameters(double A, double B, double C, double D)
{
    a = A;
    b = B;
    c = C;
    d = D;
}

double ExpressionEvaluator::getA() const { return a; }
double ExpressionEvaluator::getB() const { return b; }
double ExpressionEvaluator::getC() const { return c; }
double ExpressionEvaluator::getD() const { return d; }

double ExpressionEvaluator::Calculate() const
{
    if (d == 0)
        throw std::runtime_error("D не може бути 0");

    double logCalculate = 4 * b - c;

    MathOperation mathOps;
    double resultLog = mathOps.Logarithm(logCalculate);

    double denominator = b + (c / d) - 1;

    if (denominator == 0)
        throw std::runtime_error("Знаменник дорівнює 0");

    return (resultLog + a) / denominator;
}

std::string ExpressionEvaluator::GetValues() const
{
    return std::to_string(Calculate());
}

File: C#/Lab7.Library/ExpressionEvaluator.cs
using System;

namespace Lab7.Library
{
    public class ExpressionEvaluator
    {
        public double A { get; private set; }
        public double B { get; private set; }
        public double C { get; private set; }
        public double D { get; private set; }

        public ExpressionEvaluator(double a, double b, double c, double d)
        {
            A = a;
            B = b;
            C = c;
            D = d;
        }

        public void SetParameters(double a, double b, double c, double d)
        {
            A = a;
            B = b;
            C = c;
            D = d;
        }

        public double Calculate()
        {
            if (D == 0) 
                throw new DivideByZeroException("D не може бути нулем");

            double logCalculate = 4 * B - C;

            MathOperation mathOps = new MathOperation();
            double resultLog = mathOps.Logarithm(logCalculate);

            double denominator = B + (C / D) - 1;

            if (denominator == 0)
                throw new DivideByZeroException("Знаменник не може бути нулем");

            return (resultLog * A) / denominator;
        }

        public string GetValues()
        {
            string res = " ";
            res = res + Calculate();
            return res;
        }
    }
}